### ISSUE TITLE
Problem with running Optuna study because Lightning cannot be imported (not found)

### ISSUE THREAD

**Issue by Mballerina** [optuna/optuna#5199]
### Motivation

Mac M1 with pytorch Lightning prevents optuna study.optimize() from running without adaptation of code.

### Suggestion

Code needs to be adapted so that Lightning Trainer defaults to 'auto': 

PYTORCH_ENABLE_MPS_FALLBACK=1

import warnings

from packaging import version

import optuna
from optuna.storages._cached_storage import _CachedStorage
from optuna.storages._rdb.storage import RDBStorage

# Define key names of `Trial.system_attrs`.
_PRUNED_KEY = "ddp_pl:pruned"
_EPOCH_KEY = "ddp_pl:epoch"

with optuna._imports.try_import() as _imports:
    import pytorch_lightning as pl
    from pytorch_lightning import LightningModule
    from pytorch_lightning import Trainer   # here trainer only accepts mps and cpu, but optimization cannot handle mps always so 'auto'
    from pytorch_lightning.callbacks import Callback
    from pytorch_lightning.strategies import ParallelStrategy

if not _imports.is_successful():
    Callback = object  # type: ignore  # NOQA
    LightningModule = object  # type: ignore  # NOQA
    Trainer = object  # type: ignore  # NOQA

def _is_ddp_backend(strategy):
    """Check for DDP training strategy.

    This may be too inclusive a check, but I haven't tested it comprehensively.
    Derived from https://github.com/Lightning-AI/lightning/blob/e57f12851fe68ccfb190b368e173d5f4fad67a41/src/lightning/pytorch/trainer/connectors/accelerator_connector.py#L223
    """
    is_ddp_str = isinstance(strategy, str) and "ddp" in strategy
    is_deepspeed_str = isinstance(strategy, str) and "deepspeed" in strategy
    return isinstance(strategy, ParallelStrategy) or is_ddp_str or is_deepspeed_str

if version.parse(pl.__version__) >= version.parse("1.8.0"):

    class PyTorchLightningPruningCallback(Callback):
        """PyTorch Lightning callback to prune unpromising trials.

        See `the example <https://github.com/optuna/optuna-examples/blob/
        main/pytorch/pytorch_lightning_simple.py>`__
        if you want to add a pruning callback which observes accuracy.

        Args:
            trial:
                A :class:`~optuna.trial.Trial` corresponding to the current evaluation of the
                objective function.
            monitor:
                An evaluation metric for pruning, e.g., ``val_loss`` or
                ``val_acc``. The metrics are obtained from the returned dictionaries from e.g.
                ``pytorch_lightning.LightningModule.training_step`` or
                ``pytorch_lightning.LightningModule.validation_epoch_end`` and the names thus depend on
                how this dictionary is formatted.

        .. note::
            For the distributed data parallel training, the version of PyTorchLightning needs to be
            higher than or equal to v1.5.0. In addition, :class:`~optuna.study.Study` should be
            instantiated with RDB storage.
        """

        def __init__(self, trial: optuna.trial.Trial, monitor: str) -> None:
            _imports.check()
            super().__init__()

            self._trial = trial
            self.monitor = monitor
            self.is_ddp_backend = False

        def setup(self, trainer: Trainer(accelerator='auto'), pl_module: LightningModule, stage: str) -> None:
            # Fix
            self.is_ddp_backend = _is_ddp_backend(trainer._accelerator_connector.strategy)
            if self.is_ddp_backend:
                if not (
                    isinstance(self._trial.study._storage, _CachedStorage)
                    and isinstance(self._trial.study._storage._backend, RDBStorage)
                ):
                    raise ValueError(
                        "optuna.integration.PyTorchLightningPruningCallback"
                        " supports only optuna.storages.RDBStorage in DDP."
                    )

        def on_validation_end(self, trainer: Trainer(accelerator='auto'), pl_module: LightningModule) -> None:

            # When the trainer calls `on_validation_end` for sanity check,
            # do not call `trial.report` to avoid calling `trial.report` multiple times
            # at epoch 0. The related page is
            # https://github.com/PyTorchLightning/pytorch-lightning/issues/1391.
            if trainer.sanity_checking:
                return

            epoch = pl_module.current_epoch

            current_score = trainer.callback_metrics.get(self.monitor)
            if current_score is None:
                message = (
                    "The metric '{}' is not in the evaluation logs for pruning. "
                    "Please make sure you set the correct metric name.".format(self.monitor)
                )
                warnings.warn(message)
                return

            should_stop = False
            if trainer.is_global_zero:
                self._trial.report(current_score.item(), step=epoch)
                should_stop = self._trial.should_prune()
            should_stop = trainer.strategy.broadcast(should_stop)
            if not should_stop:
                return

            if not self.is_ddp_backend:
                message = "Trial was pruned at epoch {}.".format(epoch)
                raise optuna.TrialPruned(message)
            else:
                # Stop every DDP process if global rank 0 process decides to stop.
                trainer.should_stop = True
                if trainer.is_global_zero:
                    self._trial.storage.set_trial_system_attr(self._trial._trial_id, _PRUNED_KEY, True)
                    self._trial.storage.set_trial_system_attr(self._trial._trial_id, _EPOCH_KEY, epoch)

        def on_fit_end(self, trainer: Trainer(accelerator='auto'), pl_module: LightningModule) -> None:
            if not self.is_ddp_backend:
                return

            # Because on_validation_end is executed in spawned processes,
            # _trial.report is necessary to update the memory in main process, not to update the RDB.
            _trial_id = self._trial._trial_id
            _study = self._trial.study
            _trial = _study._storage._backend.get_trial(_trial_id)  # type: ignore
            _trial_system_attrs = _study._storage.get_trial_system_attrs(_trial_id)
            is_pruned = _trial_system_attrs.get(_PRUNED_KEY)
            epoch = _trial_system_attrs.get(_EPOCH_KEY)
            intermediate_values = _trial.intermediate_values
            for step, value in intermediate_values.items():
                self._trial.report(value, step=step)

            if is_pruned:
                message = "Trial was pruned at epoch {}.".format(epoch)
                raise optuna.TrialPruned(message)

else:
    PyTorchLightningPruningCallback = optuna.integration.PyTorchLightningPruningCallback

### Additional context (optional)

Additionally I am working with tcn the arguments need to be passed for when weight_norm (and other non mps ) functions are included;

# GPU/MPS/CPU: detect if a GPU is available
    if torch.cuda.is_available():
        pl_trainer_kwargs = {
            "accelerator": "gpu",  #used to be gpu
            "gpus": -1,
            "auto_select_gpus": True,
            "callbacks": callbacks,
            #"callbacks": [TFMProgressBar(enable_train_bar_only=True)],  #disable progress bars for all model stages except training.
        }
        num_workers = 4
    elif torch.cuda.is_available()==False:
        pl_trainer_kwargs = {
            "accelerator": "auto",  #detect if MPS can be used if no GPU
            "callbacks": callbacks}
        num_workers = 0
    else:
        pl_trainer_kwargs = {
            "accelerator": "cpu", #fallback on CPU
            "callbacks": callbacks}
        num_workers = 0

**Comment by nzw0301** [comment]
Could you fix the code snippet to run your script?

**Comment by AndrewJGroves** [comment]
Hi, I got the same problem running this example for darts 
https://github.com/unit8co/darts/blob/master/examples/17-hyperparameter-optimization.ipynb

**Comment by nzw0301** [comment]
@AndrewJGroves Could you share the minimal reproducible code? The shared notebook is too long.

**Comment by nzw0301** [comment]
Wr could not reproduce the error or get any updates from users. Let me close this issue.
